src/configuration/configuration.service.ts

```diff
@@ -0,0 +1,392 @@
+      if (existingVarIndex > -1) {
+        service.globalVariables[existingVarIndex] = variable;
+      } else {
+        service.globalVariables.push(variable);
+      }
+    });
+  }
+
+  /**
+   * Updates multiple replica configuration variables.
+   * @param serviceName - The name of the service.
+   * @param replicaId - The ID of the replica.
+   * @param variables - The updated variables.
+   * @returns The updated service configuration.
+   * @throws NotFoundException if the service or replica is not found.
+   */
+  batchAddOrUpdateReplicaVariables(
+    serviceName: string,
+    replicaId: string,
+    variables: ConfigurationVariable[],
+  ): ServiceConfiguration {
+    try {
+      const service = this.findService(serviceName);
+      if (!service) {
+        throw new NotFoundException(`Service '${serviceName}' not found`);
+      }
+      const replica = service.replicas.find(
+        (existingReplica) => existingReplica.id === replicaId,
+      );
+      if (!replica) {
+        throw new NotFoundException(`Replica '${replicaId}' not found`);
+      }
+      this.validateVariables(variables, serviceName);
+      this.updateReplicaVariables(variables, replica);
+      // send updated configuration to sidecar
+      this.eventService.publishConfiguration(serviceName, [replica]);
+      return service;
+    } catch (error) {
+      this.logger.error(`{batchAddOrUpdateReplicaVariables} ${error.message}`)
+      throw error;
+    }
+  }
+
+  /**
+   * Updates the replica variables with the given configuration variables.
+   * If a variable with the same key already exists in the replica, it will be updated.
+   * Otherwise, the variable will be added to the replica.
+   * 
+   * @param variables - The configuration variables to update the replica with.
+   * @param replica - The service replica to update.
+   */
+  private updateReplicaVariables(variables: ConfigurationVariable[], replica: ServiceReplica) {
+    variables.forEach((variable) => {
+      const existingVarIndex = replica.replicaVariables.findIndex(
+        (existingVariable) => existingVariable.key === variable.key
+      );
+      if (existingVarIndex > -1) {
+        replica.replicaVariables[existingVarIndex] = variable;
+      } else {
+        replica.replicaVariables.push(variable);
+      }
+    });
+  }
+
+  /**
+   * Validates the configuration variables against the variable definitions of a service.
+   * @param variables - The configuration variables to validate.
+   * @param serviceName - The name of the service.
+   * @throws Error if the variables do not match the variable definitions.
+   */
+  validateVariables(variables: ConfigurationVariable[], serviceName: string) {
+    const { variableDefinitions } = this.findService(serviceName);
+    variables.forEach(variable => {
+      const definition = variableDefinitions.find(def => def.key === variable.key);
+      if (!definition) throw new Error(`Variable definition not found for ${variable.key}`);
```

Comment:
brackets pls

Reply:
Changed in [2838e7949e949515df686773663b51b0a698984d](https://github.com/MiSArch/experiment-config-frontend/pull/1/commits/2838e7949e949515df686773663b51b0a698984d)